### PR TITLE
EDAC: Fixing API and API docs generation, fixing static analyzer warnings

### PR DESCRIPTION
**Pull request by finikorg**
Make it possible to generate full EDAC API for documentation.
Fixing warnings from static analyzer.

### CODE REVIEW

File: drivers/edac/edac_ibecc.c [hunk 1]
```diff
@@ -207,9 +207,11 @@ static int inject_set_param1(const struct device *dev, uint64_t addr)
 	return 0;
 }
 
-static uint64_t inject_get_param1(const struct device *dev)
+static int inject_get_param1(const struct device *dev, uint64_t *value)
```
Comment (nashif) [inline drivers/edac/edac_ibecc.c:210]: this violates misra dir 4.6: "typedefs that indicate size and signedness should be used in place of the basic numerical types"

Cane we please keep using typedefs instead.

File: include/drivers/edac.h [hunk 2]
```diff
@@ -239,29 +266,38 @@ static inline void edac_ecc_error_log_clear(const struct device *dev)
...
  */
-static inline uint64_t edac_parity_error_log_get(const struct device *dev)
+static inline int edac_parity_error_log_get(const struct device *dev,
+					    uint64_t *value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->parity_error_log_get(dev);
+	__ASSERT(api->parity_error_log_get,
```
Comment (nashif) [inline include/drivers/edac.h:285]: this should return -ENOSYS, see https://docs.zephyrproject.org/latest/guides/design_guidelines.html#return-codes

**Review by nashif — DISMISSED** [review]

File: drivers/edac/edac_ibecc.c (hunk 1 above)
Comment (finikorg) [reply drivers/edac/edac_ibecc.c:210]: @nashif , sorry do not get this. with respect to this change above typedef does not matter. I return here `int` instead of value to be able to return `-ENOTSUP`.

File: include/drivers/edac.h (hunk 2 above)
Comment (finikorg) [reply include/drivers/edac.h]: Updated in 65f0dc2

File: drivers/edac/edac_ibecc.c (hunk 1 above)
Comment (jhedberg) [reply drivers/edac/edac_ibecc.c:210]: If the intent is to return `0` on success and `-errno` on failure then `int` is the appropriate choice, I believe.

File: drivers/edac/shell.c [hunk 3]
```diff
@@ -62,6 +62,7 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
 	uint64_t error;
+	int ret;
```
Comment (jhedberg) [inline drivers/edac/shell.c:65]: Please use `err` for simple `int` type errors. If you think this will be confusing with your `error` variable, then rename that one e.g. to `ecc_err` or something similar.

File: drivers/edac/shell.c [hunk 4]
```diff
@@ -71,27 +72,38 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
...
 
-	return 0;
+	return ret;
 }
 
 #if defined(CONFIG_EDAC_ERROR_INJECT)
 static int cmd_inject_addr(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
+	int ret;
```
Comment (jhedberg) [inline drivers/edac/shell.c:106]: Same here

File: drivers/edac/shell.c [hunk 5]
```diff
@@ -116,15 +134,21 @@ static int cmd_inject_addr(const struct shell *shell, size_t argc, char **argv)
...
 	}
 
-	return 0;
+	return ret;
 }
 
 static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
+	int ret;
```
Comment (jhedberg) [inline drivers/edac/shell.c:151]: And here

File: drivers/edac/shell.c [hunk 6]
```diff
@@ -71,27 +72,38 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
 
 	shell_fprintf(shell, SHELL_NORMAL, "Show EDAC status\n");
 
-	error = edac_ecc_error_log_get(dev);
+	ret = edac_ecc_error_log_get(dev, &error);
+	if (ret) {
```
Comment (jhedberg) [inline drivers/edac/shell.c:76]: Assuming this is under the scope that the coding style will be enforced on, I think only boolean expressions or bool type variables can be used as if-conditions, so this would have to be `if (ret != 0)`. @nashif can confirm.

File: drivers/edac/edac_ibecc.c (hunk 1 above)
Comment (nashif) [reply drivers/edac/edac_ibecc.c:210]: > If the intent is to return `0` on success and `-errno` on failure then `int` is the appropriate choice, I believe.

directive 4.6 says otherwise though.

Comment (jhedberg) [reply drivers/edac/edac_ibecc.c:210]: I guess we might need to state an explicit exception to this rule in our coding style conventions then, since I believe `int` return is used in lots of places of Zephyr, in particular in the core kernel code. If I instead see e.g. `int32_t` I will think that the return value has some other meaning than the usual "0 or -errno", whereas if I see a `int` return value I can assume that it follows the posix error number semantics. I think this is a valuable convention to keep around since that way we keep it separate from some use cases that really do require a fixed-width integer.

Comment (finikorg) [reply drivers/edac/edac_ibecc.c:210]: @nashif I did not understood first that you are talking about this change `int` -> `int32_t`. I thought the comment was about removing some typedefs above. I can change `int` to `int32_t`, although, as @jhedberg mentioned, this is (was?) common for return type in Zephyr API.

File: drivers/edac/shell.c (hunk 6 above)
Comment (finikorg) [reply drivers/edac/shell.c:76]: Ok, I see @nashif reaction, will change to boolean.

File: drivers/edac/shell.c (hunk 3 above)
Comment (finikorg) [reply drivers/edac/shell.c]: done

File: drivers/edac/edac_ibecc.c [hunk 7]
```diff
@@ -371,7 +386,7 @@ int edac_ibecc_init(const struct device *dev)
 		return -ENODEV;
 	}
 
-	if (!ibecc_enabled(bdf)) {
+	if (ibecc_enabled(bdf) == false) {
```
Comment (jhedberg) [inline drivers/edac/edac_ibecc.c:389]: If `ibecc_enabled()` returns `bool` then I don't think you need to do this modification (the bool type is a valid boolean statement in itself)

File: include/drivers/edac.h [hunk 8]
```diff
@@ -14,93 +14,77 @@
...
  */
 static inline int edac_inject_set_param1(const struct device *dev,
-					 uint64_t addr)
+					 uint64_t value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_set_param1(dev, addr);
+	if (!api->inject_set_param1) {
```
Comment (jhedberg) [inline include/drivers/edac.h:83]: `if (api->inject_set_param == NULL)`?

File: include/drivers/edac.h [hunk 9]
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
...
  */
-static inline uint64_t edac_inject_get_param1(const struct device *dev)
+static inline int edac_inject_get_param1(const struct device *dev,
+					 uint64_t *value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_get_param1(dev);
+	if (!api->inject_get_param1) {
```
Comment (jhedberg) [inline include/drivers/edac.h:107]: `if (api->inject_set_param == NULL)`?

File: include/drivers/edac.h [hunk 10]
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
...
  */
 static inline int edac_inject_set_param2(const struct device *dev,
-					 uint64_t mask)
+					 uint64_t value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_set_param2(dev, mask);
+	if (!api->inject_set_param2) {
```
Comment (jhedberg) [inline include/drivers/edac.h:132]: `if (api->inject_set_param == NULL)`?

File: include/drivers/edac.h [hunk 11]
```diff
@@ -160,14 +165,20 @@ static inline uint64_t edac_inject_get_param2(const struct device *dev)
...
+ * @retval -ENOTSUP if not supported
+ * @retval 0 on success, error code otherwise
  */
 static inline int edac_inject_set_error_type(const struct device *dev,
 					     uint32_t error_type)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->inject_set_error_type) {
```
Comment (jhedberg) [inline include/drivers/edac.h:178]: `if (api->inject_set_error_type == NULL)`?

File: include/drivers/edac.h [hunk 12]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
+ *
+ * @retval -ENOTSUP if not supported
+ * @retval 0 on success, error code otherwise
  */
 static inline int edac_inject_error_trigger(const struct device *dev)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->inject_error_trigger) {
```
Comment (jhedberg) [inline include/drivers/edac.h:224]: Same here

File: include/drivers/edac.h [hunk 13]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
  */
-static inline uint64_t edac_ecc_error_log_get(const struct device *dev)
+static inline int edac_ecc_error_log_get(const struct device *dev,
+					 uint64_t *value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->ecc_error_log_get(dev);
+	if (!api->ecc_error_log_get) {
```
Comment (jhedberg) [inline include/drivers/edac.h:250]: And here

File: include/drivers/edac.h [hunk 14]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
+ * @retval -ENOSYS if the mandatory interface is not implemented
  */
-static inline void edac_ecc_error_log_clear(const struct device *dev)
+static inline int edac_ecc_error_log_clear(const struct device *dev)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	api->ecc_error_log_clear(dev);
+	if (!api->ecc_error_log_clear) {
```
Comment (jhedberg) [inline include/drivers/edac.h:272]: And here

File: include/drivers/edac.h [hunk 15]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
  */
-static inline uint64_t edac_parity_error_log_get(const struct device *dev)
+static inline int edac_parity_error_log_get(const struct device *dev,
+					    uint64_t *value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->parity_error_log_get(dev);
+	if (!api->parity_error_log_get) {
```
Comment (jhedberg) [inline include/drivers/edac.h:296]: And here

File: include/drivers/edac.h [hunk 16]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
+ * @retval -ENOSYS if the mandatory interface is not implemented
  */
-static inline void edac_parity_error_log_clear(const struct device *dev)
+static inline int edac_parity_error_log_clear(const struct device *dev)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	api->parity_error_log_clear(dev);
+	if (!api->parity_error_log_clear) {
```
Comment (jhedberg) [inline include/drivers/edac.h:318]: And here

File: include/drivers/edac.h [hunk 17]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
+ * @retval num Number of correctable errors
+ * @retval -ENOSYS if the mandatory interface is not implemented
  */
-static inline unsigned int edac_errors_cor_get(const struct device *dev)
+static inline int edac_errors_cor_get(const struct device *dev)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->errors_cor_get) {
```
Comment (jhedberg) [inline include/drivers/edac.h:338]: And here

File: include/drivers/edac.h [hunk 18]
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
...
+ * @retval num Number of uncorrectable errors
+ * @retval -ENOSYS if the mandatory interface is not implemented
  */
-static inline unsigned int edac_errors_uc_get(const struct device *dev)
+static inline int edac_errors_uc_get(const struct device *dev)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->errors_uc_get) {
```
Comment (jhedberg) [inline include/drivers/edac.h:358]: And here

File: include/drivers/edac.h [hunk 19]
```diff
@@ -303,13 +369,19 @@ static inline unsigned int edac_errors_uc_get(const struct device *dev)
...
+ *
+ * @retval 0 on success, error code otherwise
+ * @retval -ENOSYS if the mandatory interface is not implemented
  */
 static inline int edac_notify_callback_set(const struct device *dev,
 					   edac_notify_callback_f cb)
 {
 	const struct edac_driver_api *api = dev->api;
 
+	if (!api->notify_cb_set) {
```
Comment (jhedberg) [inline include/drivers/edac.h:381]: And here

File: include/drivers/edac.h [hunk 20]
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
...
-static inline uint64_t edac_inject_get_param1(const struct device *dev)
+static inline int edac_inject_get_param1(const struct device *dev,
+					 uint64_t *value)
 {
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_get_param1(dev);
+	if (api->inject_get_param1 == NULL) {
+		return -ENOTSUP;
```
Comment (jhedberg) [inline include/drivers/edac.h:108]: Sorry, I should have noticed this detail earlier. Whenever an API isn't implemented the right error is `ENOSYS` and not `ENOTSUP`. See here: https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/api/design_guidelines.rst#return-codes

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: include/drivers/edac.h (hunk 20 above)
Comment (finikorg) [reply include/drivers/edac.h:108]: @jhedberg is `ENOSYS` is for mandatory API? This is optional API.

Comment (jhedberg) [reply include/drivers/edac.h:108]: @finikorg I don't think it makes any difference. Grep for ENOSYS under include/drivers to see how it's used.

File: drivers/edac/edac_ibecc.c (hunk 1 above)
Comment (finikorg) [reply drivers/edac/edac_ibecc.c]: @nashif Added 617c528  converting `int` to `int32_t`.

File: include/drivers/edac.h [hunk 21]
```diff
@@ -14,93 +14,80 @@
...
- * @param addr Injection address base
- * @return 0 on success, error code otherwise
+ * @param value First injection parameter
+ *
+ * @retval -ENOSYS if the optional interface is not implemented
+ * @retval 0 on success, other error code otherwise
  */
-static inline int edac_inject_set_param1(const struct device *dev,
-					 uint64_t addr)
+static inline int32_t edac_inject_set_param1(const struct device *dev,
```
Comment (jhedberg) [inline include/drivers/edac.h:80]: The convention for APIs that return negative POSIX error numbers is `int` and not `int32`.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: include/drivers/edac.h (hunk 21 above)
Comment (nashif) [reply include/drivers/edac.h:80]: what convention exactly? this convention you are talking about conflicts with the coding guidelines adopted by Zephyr.

Comment (jhedberg) [reply include/drivers/edac.h:80]: I mean't a de-facto convention, i.e. an established practice that can be seen throughout the tree. Whenever you see an `int` return you can assume that it has these semantics. I just counted 66 such APIs in `include/kernel.h` alone. If we move away from `int` I don't think `int32_t` is the right choice, since it looses essential semantic information  IMO. Instead we could introduce some new `err_t` type which is known to have these semantics.

Comment (jhedberg) [reply include/drivers/edac.h:80]: @finikorg I think you're referencing the wrong commit. In general, I think this has project-wide implications so review comments in this PR aren't necessary the right place for the discussion. I'd consider the API or Dev meetings instead.

Comment (finikorg) [reply include/drivers/edac.h:80]: @jhedberg Fixed commit.

Comment (carlescufi) [reply include/drivers/edac.h:80]: This has been raised before. I am of the opinion that using `int` wherever the size in bits of the variable doesn't matter is the correct approach, because it uses the native integer type of the platform. However, if this gets in the way of safety certification I agree we should consider it, and slowly transition to `int32_t` everywhere. In any case the error codes should always be documented with Doxygen so that it is clear for the user what to expect as a return value.
So to summarize my position: I'd rather use `int`, but safety certification is more important than my personal taste. If safety requires this then we need to transition to using `int32_t` everywhere for consistency, I assume then also on 64-bit platforms.

Comment (carlescufi) [reply include/drivers/edac.h:80]: @nashif wouldn't `size_t` be an issue as well in this case? since its size is platform-dependant, as @jhedberg observed offline.

Comment (jhedberg) [reply include/drivers/edac.h:80]: Btw, this also relates to #32213 , i.e. if we are going to change the commonly used error type we might want to introduce the concept of domain in the same go.

File: drivers/edac/edac_ibecc.c (hunk 1 above)
Comment (sjg20) [reply drivers/edac/edac_ibecc.c]: IMO this line is correct. We should use int (i.e. the natural size of the machine) for return codes, not int32_t or int64_t

File: include/drivers/edac.h (hunk 21 above)
Comment (sjg20) [reply include/drivers/edac.h:80]: I was looking here. Is it true that the rule has become mandatory in the 8 years from 2004 to 2012?

https://bullwhip.physio-control.com/documentation/help/reference/misra.builtin_numeric.htm

I agree with carlescufi on this point. 

Also, what do we do on 64-bit platforms with these APIs? Disallowing native types for error returns seems like a strange rule to me. You cannot stop an int32_t being promoted to a 64-bit value if that is the machine's register size. It might just add more confusion.

The page referenced above refers to variables that store some value with a meaning (speed, torque). I am not even sure that error codes fit that definition. If they do end up being stored in a data structure then using a specific length perhaps makes more sense.

Comment (finikorg) [reply include/drivers/edac.h]: I removed that patch introducing that change `int` -> `int32_t`. I did not like it at all, just had no other option ;)

**Comment by finikorg** [comment]
Rebased against latest upstream

File: drivers/edac/shell.c [hunk 22]
```diff
@@ -140,7 +164,13 @@ static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
 	}
 
 	if (argc == 1) {
-		uint64_t mask = edac_inject_get_param2(dev);
+		uint64_t mask;
+
+		err = edac_inject_get_param2(dev, &mask);
+		if (err != 0) {
+			shell_error(shell, "Error getting mask");
```
Comment (jhedberg) [inline drivers/edac/shell.c:171]: Maybe print the exact error as well with %d?

File: drivers/edac/shell.c [hunk 23]
```diff
@@ -150,10 +180,14 @@ static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
 		shell_fprintf(shell, SHELL_NORMAL,
 			      "Set injection address mask to %lx\n", value);
 
-		return edac_inject_set_param2(dev, value);
+		err = edac_inject_set_param2(dev, value);
+		if (err != 0) {
+			shell_error(shell, "Error setting mask");
```
Comment (jhedberg) [inline drivers/edac/shell.c:185]: Same here

File: drivers/edac/shell.c (hunk 22 above)
Comment (jhedberg) [reply drivers/edac/shell.c:171]: I actually noticed that it's missing from many existing error cases, so a general improvement you could do as a separate patch

File: drivers/edac/shell.c [hunk 24]
```diff
@@ -62,39 +62,51 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
...
 
-	if (error) {
+	if (error != 0) {
 		decode_ecc_error(shell, error);
 	}
 
-	error = edac_parity_error_log_get(dev);
+	err = edac_parity_error_log_get(dev, &error);
+	if (err != 0) {
+		shell_error(shell, "Error getting parity error log");
```
Comment (jhedberg) [inline drivers/edac/shell.c:89]: Missing err %d

File: drivers/edac/shell.c [hunk 25]
```diff
@@ -62,39 +62,51 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
...
 		shell_error(shell, "IBECC device not found");
 		return -ENODEV;
 	}
 
 	shell_fprintf(shell, SHELL_NORMAL, "Show EDAC status\n");
 
-	error = edac_ecc_error_log_get(dev);
+	err = edac_ecc_error_log_get(dev, &error);
+	if (err != 0) {
+		shell_error(shell, "Error getting ecc error log");
```
Comment (jhedberg) [inline drivers/edac/shell.c:77]: Missing err %d

**Review by jhedberg — APPROVED** [review]
Looks fine, except for a couple of places with missing `err %d`

File: drivers/edac/shell.c (hunk 24 above)
Comment (finikorg) [reply drivers/edac/shell.c]: Sorry, missed these, fixed now

**Review by nashif — APPROVED** [review]
